main/src/library/Chain.flix

```diff
@@ -0,0 +1,847 @@
+        case SomeLeft(x, rs) => reverseHelper(rs, cons(x, ac))
+    }
+
+    ///
+    /// Returns `xs` with `a` inserted between every two adjacent elements.
+    ///
+    pub def intersperse(a: a, xs: Chain[a]): Chain[a] = match viewLeft(xs) {
+        case NoneLeft        => Empty
+        case SomeLeft(x, rs) => intersperseHelper(a, rs, point(x))
+    }
+
+    ///
+    /// Helper function for `reverse`.
+    ///
+    def intersperseHelper(sep: a, xs: Chain[a], ac: Chain[a]): Chain[a] = match viewLeft(xs) {
+        case NoneLeft        => ac
+        case SomeLeft(x, rs) => intersperseHelper(sep, rs, ac `snoc` sep `snoc` x)
+    }
+
+    ///
+    /// Returns the number of elements in `xs` that satisfy the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def count(f: a -> Bool, xs: Chain[a]): Int32 = countHelper(f, xs, 0)
+
+    ///
+    /// Helper function for `count`.
+    ///
+    def countHelper(f: a -> Bool, xs: Chain[a], ac: Int32): Int32 = match viewLeft(xs) {
+        case NoneLeft        => ac
+        case SomeLeft(x, rs) => if (f(x)) countHelper(f, rs, ac + 1) else countHelper(f, rs, ac)
+    }
+
+    ///
+    /// Returns the concatenation of the elements in `xs`.
+    ///
+    pub def flatten(xs: Chain[Chain[a]]): Chain[a] =
+        foldLeft(append, empty(), xs)
+
+    ///
+    /// Returns `true` if and only if at least one element in `xs` satisfies the predicate `f`.
+    ///
+    /// Returns `false` if `xs` is empty.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def exists(f: a -> Bool, xs: Chain[a]): Bool = match viewLeft(xs) {
+        case NoneLeft        => false
+        case SomeLeft(x, rs) => if (f(x)) true else exists(f, rs)
+    }
+
+    ///
+    /// Returns `true` if and only if all elements in `xs` satisfy the predicate `f`.
+    ///
+    /// Returns `true` if `xs` is empty.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def forall(f: a -> Bool, xs: Chain[a]): Bool = match viewLeft(xs) {
+        case NoneLeft        => true
+        case SomeLeft(x, rs) => if (not f(x)) false else forall(f, rs)
+    }
+
+    ///
+    /// Returns a list of every element in `xs` that satisfies the predicate `f`.
+    ///
+    /// The function `f` must be pure.
+    ///
+    pub def filter(f: a -> Bool & ef, xs: Chain[a]): Chain[a] & ef = filterHelper(f, xs, Empty)
+
+    ///
+    /// Helper function for `filter`.
+    ///
+    def filterHelper(f: a -> Bool & ef, xs: Chain[a], ac: Chain[a]): Chain[a] & ef = match viewLeft(xs) {
+        case NoneLeft        => ac
+        case SomeLeft(x, rs) => if (f(x)) filterHelper(f, rs, snoc(ac, x)) else filterHelper(f, rs, ac)
+    }
+
+    ///
+    /// Applies `f` to a start value `s` and all elements in `xs` going from left to right.
+    ///
+    /// That is, the result is of the form: `f(...f(f(s, x1), x2)..., xn)`.
+    ///
+    pub def foldLeft(f: (b, a) -> b & ef, ac: b, xs: Chain[a]): b & ef = match viewLeft(xs) {
```

Comment:
I think this reveals the hand: When you use viewLeft you must be seeing the elements from left-to-right. 

Nevertheless, some doc further up would be useful! :)

Reply:
There are more comments on both the exported functions `viewLeft` & `viewRight` and their helper functions.

The comments on the helpers make it explicit that you are traversing left-to-right or right-to-left.